os-boot/linux/Makefile

```
@@ -0,0 +1,100 @@
+##### Compilation #####
+
+RISCV_COMPILER_PREFIX := riscv64-buildroot-linux-gnu-
+
+# This is the final bootable OpenSBI + Linux image.
+elf := build/fw_payload.elf
+
+.PHONY: all
+all: $(elf)
+
+# Generic rule to download tools.
+downloads/%.tar.xz: %.url
+	mkdir -p downloads
+	curl --location '$(shell cat $<)' --output $@
+
+# Rules to extract tools. For simplicity we strip the first directory component because it varies.
+build/gcc/bin/$(RISCV_COMPILER_PREFIX)gcc: downloads/gcc.tar.xz
+	mkdir -p build/gcc
+	tar --touch --directory build/gcc --strip-components=1 --extract --file downloads/gcc.tar.xz
+
+build/linux/Makefile: downloads/linux.tar.xz
+	mkdir -p build/linux
+	tar --touch --directory build/linux --strip-components=1 --extract --file downloads/linux.tar.xz
+
+build/opensbi/Makefile: downloads/opensbi.tar.xz
+	mkdir -p build/opensbi
+	tar --touch --directory build/opensbi --strip-components=1 --extract --file downloads/opensbi.tar.xz
+
+CROSS_COMPILE := $(shell pwd)/build/gcc/bin/$(RISCV_COMPILER_PREFIX)
+
+# Rule to build the Linux kernel.
+build/linux/arch/riscv/boot/Image: build/linux/Makefile build/gcc/bin/$(RISCV_COMPILER_PREFIX)gcc
+	$(MAKE) -C build/linux CROSS_COMPILE=$(CROSS_COMPILE) ARCH=riscv defconfig
+	$(MAKE) -C build/linux CROSS_COMPILE=$(CROSS_COMPILE) CONFIG_HVC_RISCV_SBI=y ARCH=riscv Image
+
+# Rule to build OpenSBI, with the Linux kernel embedded in it.
+#
+# FW_TEXT_START is 0 by default which doesn't leave space for the emulator bootloader.
+# 0x80000000 is the default start of Spike's memory.
+$(elf): build/opensbi/Makefile build/linux/arch/riscv/boot/Image build/gcc/bin/$(RISCV_COMPILER_PREFIX)gcc
+	$(MAKE) -C build/opensbi FW_TEXT_START=0x80000000 FW_PAYLOAD=y FW_PAYLOAD_PATH=../linux/arch/riscv/boot/Image PLATFORM=generic CROSS_COMPILE=$(CROSS_COMPILE)
+	cp build/opensbi/build/platform/generic/firmware/fw_payload.elf $(elf)
+
+build/sail.dts:
```

```suggestion
build/sail.dts: ../../build/c_emulator/riscv_sim_rv64d
```
Shouldn't this be re-run when the sail simulator changes?